File: js_package/everyai-core/src/cache.js
import crypto from "crypto";
import fs from "fs";
import path from "path";
import os from "os";

/**
 * Handles local response caching to prevent duplicate API execution.
 */
export class RequestCache {
    /**
     * @param {string} [cachePath] Optional custom file path for the cache.
     */
    constructor(cachePath = null) {
        if (cachePath === null) {
            const dbDir = path.join(os.homedir(), ".everyai");
            if (!fs.existsSync(dbDir)) {
                try {
                    fs.mkdirSync(dbDir, { recursive: true });
                } catch (e) {
                    // Fallback to in-memory if home directory is not writable
                }
            }
            this.cachePath = path.join(dbDir, "cache.json");
        } else {
            this.cachePath = cachePath;
        }

        this.memoryCache = {};
        this._loadCache();
    }

    /**
     * Load cache from disk into memory.
     * @private
     */
    _loadCache() {
        if (fs.existsSync(this.cachePath)) {
            try {
                const data = fs.readFileSync(this.cachePath, "utf-8");
                this.memoryCache = JSON.parse(data);
            } catch (e) {
                this.memoryCache = {};
            }
        }
    }

    /**
     * Save cache from memory to disk.
     * @private
     */
    _saveCache() {
        try {
            const dir = path.dirname(this.cachePath);
            if (!fs.existsSync(dir)) {
                fs.mkdirSync(dir, { recursive: true });
            }
            fs.writeFileSync(this.cachePath, JSON.stringify(this.memoryCache, null, 2), "utf-8");
        } catch (e) {
            // Keep in memory if saving fails
        }
    }

    /**
     * Generate a unique SHA-256 signature for the request inputs.
     * 
     * @param {Array<Object>} messages 
     * @param {string|null} provider 
     * @param {string|null} model 
     * @param {number} temperature 
     * @param {number|null} maxTokens 
     * @param {Object} [kwargs] 
     * @returns {string} The SHA-256 hash.
     */
    _generateHash(messages, provider, model, temperature, maxTokens, kwargs = {}) {
        // Ensure message dictionaries are ordered consistently for hashing
        const normalizedMessages = messages.map(msg => {
            const sorted = {};
            Object.keys(msg).sort().forEach(key => {
                sorted[key] = msg[key];
            });
            return sorted;
        });

        const sortedKwargs = {};
        Object.keys(kwargs).sort().forEach(key => {
            sortedKwargs[key] = kwargs[key];
        });

        const payload = {
            messages: normalizedMessages,
            provider: provider ? provider.trim().toLowerCase() : null,
            model: model ? model.trim().toLowerCase() : null,
            temperature,
            max_tokens: maxTokens,
            extra_kwargs: sortedKwargs
        };

        const payloadString = JSON.stringify(payload);
        return crypto.createHash("sha256").update(payloadString, "utf8").digest("hex");
    }

    /**
     * Query the cache for a matching response.
     * 
     * @param {Array<Object>} messages 
     * @param {string|null} provider 
     * @param {string|null} model 
     * @param {number} temperature 
     * @param {number|null} maxTokens 
     * @param {Object} [kwargs] 
     * @returns {Object|null} The cached ChatCompletionResponse or null.
     */
    get(messages, provider, model, temperature, maxTokens, kwargs = {}) {
        const hash = this._generateHash(messages, provider, model, temperature, maxTokens, kwargs);
        const cached = this.memoryCache[hash];

        if (!cached) {
            return null;
        }

        // Return a copy to avoid mutation side-effects
        return JSON.parse(JSON.stringify(cached));
    }

    /**
     * Write a request/response entry to the cache.
     * 
     * @param {Array<Object>} messages 
     * @param {string|null} provider 
     * @param {string|null} model 
     * @param {number} temperature 
     * @param {number|null} maxTokens 
     * @param {Object} response The ChatCompletionResponse.
     * @param {Object} [kwargs] 
     */
    set(messages, provider, model, temperature, maxTokens, response, kwargs = {}) {
        const hash = this._generateHash(messages, provider, model, temperature, maxTokens, kwargs);
        
        // Ensure response is serialized correctly
        const serializedResponse = typeof response.toDict === "function" ? response.toDict() : response;
        
        this.memoryCache[hash] = serializedResponse;
        this._saveCache();
    }

    /**
     * Empty the cache.
     */
    clear() {
        this.memoryCache = {};
        try {
            if (fs.existsSync(this.cachePath)) {
                fs.unlinkSync(this.cachePath);
            }
        } catch (e) {
            // Ignore error
        }
    }
}

File: py_package/everyai_core/cache.py
"""Local response caching system for everyai-core.

Stores prompt-response history in a local SQLite database, allowing duplicate
requests to be resolved instantly without using tokens or triggering rate limits.
"""

import hashlib
import json
import sqlite3
from pathlib import Path
from typing import Any
from everyai_core.types import ChatCompletionResponse, ChatCompletionChoice, UsageInfo


class RequestCache:
    """Handles local response caching to prevent duplicate API execution."""

    def __init__(self, db_path: str | Path | None = None):
        """Initialize the database connection and cache table."""
        if db_path is None:
            db_dir = Path.home() / ".everyai"
            db_dir.mkdir(parents=True, exist_ok=True)
            self.db_path = db_dir / "cache.db"
        else:
            self.db_path = Path(db_path)

        self._init_db()

    def _get_connection(self) -> sqlite3.Connection:
        """Establish connection to SQLite."""
        return sqlite3.connect(self.db_path)

    def _init_db(self) -> None:
        """Initialize caching table schema."""
        with self._get_connection() as conn:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS response_cache (
                    prompt_hash TEXT PRIMARY KEY,
                    messages_json TEXT NOT NULL,
                    provider TEXT,
                    model TEXT,
                    response_json TEXT NOT NULL,
                    timestamp DATETIME DEFAULT CURRENT_TIMESTAMP
                )
            """)
            conn.commit()

    def _generate_hash(
        self,
        messages: list[dict[str, Any]],
        provider: str | None,
        model: str | None,
        temperature: float,
        max_tokens: int | None,
        **kwargs: Any
    ) -> str:
        """Generate a unique SHA-256 signature for the request inputs."""
        # Ensure message dictionaries are ordered consistently for hashing
        normalized_messages = []
        for msg in messages:
            normalized_messages.append({k: msg[k] for k in sorted(msg.keys())})

        payload = {
            "messages": normalized_messages,
            "provider": provider.strip().lower() if provider else None,
            "model": model.strip().lower() if model else None,
            "temperature": temperature,
            "max_tokens": max_tokens,
            "extra_kwargs": {k: kwargs[k] for k in sorted(kwargs.keys())}
        }
        
        payload_bytes = json.dumps(payload, sort_keys=True).encode("utf-8")
        return hashlib.sha256(payload_bytes).hexdigest()

    def get(
        self,
        messages: list[dict[str, Any]],
        provider: str | None,
        model: str | None,
        temperature: float,
        max_tokens: int | None,
        **kwargs: Any
    ) -> ChatCompletionResponse | None:
        """Query the cache for a matching response.

        Returns:
            A ChatCompletionResponse instance if found, or None.
        """
        prompt_hash = self._generate_hash(messages, provider, model, temperature, max_tokens, **kwargs)

        with self._get_connection() as conn:
            row = conn.execute(
                "SELECT response_json FROM response_cache WHERE prompt_hash = ?",
                (prompt_hash,)
            ).fetchone()

            if not row:
                return None

            try:
                data = json.loads(row[0])
                
                # Reconstruct choices dataclass items
                choices = [
                    ChatCompletionChoice(
                        index=c["index"],
                        message=c["message"],
                        finish_reason=c.get("finish_reason")
                    ) for c in data.get("choices", [])
                ]
                
                # Reconstruct token usage stats
                usage = None
                if "usage" in data and data["usage"]:
                    u = data["usage"]
                    usage = UsageInfo(
                        prompt_tokens=u.get("prompt_tokens"),
                        completion_tokens=u.get("completion_tokens"),
                        total_tokens=u.get("total_tokens")
                    )

                return ChatCompletionResponse(
                    id=data.get("id"),
                    object=data.get("object", "chat.completion"),
                    created=data.get("created", 0),
                    model=data.get("model", ""),
                    choices=choices,
                    usage=usage,
                    provider=data.get("provider")
                )
            except Exception:
                # If cache is malformed, treat it as a miss
                return None

    def set(
        self,
        messages: list[dict[str, Any]],
        provider: str | None,
        model: str | None,
        temperature: float,
        max_tokens: int | None,
        response: ChatCompletionResponse,
        **kwargs: Any
    ) -> None:
        """Write a request/response entry to the database cache."""
        prompt_hash = self._generate_hash(messages, provider, model, temperature, max_tokens, **kwargs)
        
        # Serialize standard completion dataclass
        response_dict = response.to_dict()
        response_json = json.dumps(response_dict)
        messages_json = json.dumps(messages)

        with self._get_connection() as conn:
            conn.execute(
                """
                INSERT OR REPLACE INTO response_cache 
                (prompt_hash, messages_json, provider, model, response_json, timestamp)
                VALUES (?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
                """,
                (prompt_hash, messages_json, provider, model, response_json)
            )
            conn.commit()

    def clear(self) -> None:
        """Empty the database cache."""
        with self._get_connection() as conn:
            conn.execute("DELETE FROM response_cache")
            conn.commit()
